File: IndoorLocalization/app/src/main/java/edu/onbasli/indoorlocalization/InertiaNavegation/floorDetection/FloorDetection.java
package edu.onbasli.indoorlocalization.InertiaNavegation.floorDetection;

import android.app.Activity;
import android.content.pm.PackageManager;
import android.util.Log;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;

import java.io.IOException;
import java.util.Vector;

import edu.onbasli.indoorlocalization.InertiaNavegation.filewriting.DataFileWriter;

public class FloorDetection {

    private static final String FOLDER_NAME = "Pedestrian_Dead_Reckoning/Floor_detect";
    private static final String[] DATA_FILE_NAMES = {
            "Floor_Detect"
    };
    private static final String[] DATA_FILE_HEADINGS = {
            "Barometer,currentAvg,AvgBefore_2_Second,Pstart,Pend,CurrentFloor"
    };
    private DataFileWriter dataFileWriter;

    //height estimation
    // private final int timeToUpdate = 15; //time which after we will see if the floor number changed
    private final int timeToAverage = 5; //Take avg readings every 1s
    private Vector<Float> avgReadings;
    private float avgT_2 = 0; //average values before 3 seconds;
    private float avgT_0 = 0; //average values in the last 1 seconds;
    private float avgT_5 = 0; //average values in the last 5 seconds;
    private float pstart = 0; //p at start walking up or down;
    private float pend = 0; //p at the end of walking up or down;

    private int currentFloor = 0;
    // private float lastAverageTime, lastUpdateTime = 0;
    private final float mPh_btwFloors = 0.30f;
    private boolean first_run = true;
    private boolean firstRun = true;
    private TextView floorNum;
    private float currentTime = 0;
    private final float thetaT = 0.01f; //hpa
    private int N0 , N1 = 5; //to remove ping-pong effect
    private int num1, num2 = 0; // counters
    private boolean enterLoop2 = false;
    private int upORdown = 0; //1 means up 2 means down
    private final float h0 = 3.52f; //height of each layer in the building
    private final float temperature = 27.0f;
    private final float segma = 1/273.15f;
    private final float eqConst = 18410.183f;
    private boolean areFilesCreated;


    public FloorDetection()
    {
        areFilesCreated = false;
        avgReadings = new Vector<Float>();
        createFiles();
    }

    public int GetFloorNum (float pressure) {

        if(avgReadings.size() < timeToAverage)
        {
            avgReadings.add(pressure); //pt
        }
        else {
            avgT_0 = 0;
            for (float value :avgReadings) {
                avgT_0 += value;
            }
            avgT_0 /= (float) avgReadings.size(); //pt average
            if(firstRun)
            {
                avgT_2 = avgT_0;
                firstRun = false;
            }
            avgReadings.clear();
            if(currentTime == 2)
            {
                if (!enterLoop2 ) {
                    if (Math.abs(avgT_0 - avgT_2) > thetaT) {
                        if (num1 == 0)
                            avgT_5 = avgT_0;
                        if (num1 == N0) {
                            pstart = avgT_5;
                            enterLoop2 = true;
                            num1 = 0;
                        }
                        else
                            ++num1;
                    } else {
                        num1 = 0;
                    }
                }else {
                    if (Math.abs(avgT_0 - avgT_2) < thetaT){
                        if (num2 == 0)
                            avgT_5 = avgT_0;
                        if (num2 == N1) {
                            pend = avgT_5;
                            enterLoop2 = false;
                            num2 = 0;
                            currentFloor += Math.round((1/h0) * eqConst * (1+segma * temperature) * Math.log10(pstart/pend));
                        }
                        else
                            ++num2;
                    }else {
                        num2 = 0;
                    }
                }
                avgT_2 = avgT_0;
                currentTime = 0;
            }
            ++currentTime;
        }
        dataFileWriter.writeToFile("Floor_Detect",
                (float) pressure,
                avgT_0,
                avgT_2,
                pstart,
                pend,
                (float) currentFloor);
        return 0;
    }
    private void createFiles() {
        if (!areFilesCreated) {
            try {
                dataFileWriter = new DataFileWriter(FOLDER_NAME, DATA_FILE_NAMES, DATA_FILE_HEADINGS);
            } catch (IOException e) {
                Log.e("SensorActivity", e.toString());
            }
            areFilesCreated = true;
        }
    }
}

File: Results/Height Estimation/FloorDetection.m
clear
clc
opts = detectImportOptions('Floor_Detect_2020-8-19_14-08-02.csv');
M2 = readmatrix('Floor_Detect_2020-8-19_14-08-02.csv',opts);
opts = detectImportOptions('Floor_Detect_2020-8-19_14-05-52.csv');
M1 = readmatrix('Floor_Detect_2020-8-19_14-05-52.csv',opts);
input_data = [M1(:,1)' M2(:,1)']';
avg_readings = 0;
avgT_0 = 0;
avgT_2 = 0;
enterLoop2 = false;
currentTime = 0;
h0 = 3.52;
temperature = 27.0;
segma = 1/273.15;
eqConst = 18410.183;
thetaT = 0.01;
N1 = 0; 
N0 = 5; 
num2 = 0;
num1 = 0;
firstRun = true;
currentFloor = 0;
allFloors = 0;
pstart = 0;
pend = 0;
prevI = 1;
output_file(:,1) = input_data;
output_file(:,2) = 0;
avg_countHelper = 0;
for i = 1:length(input_data)
    avg_readings(abs(i - avg_countHelper)) = input_data(i);
    if(length(avg_readings) >= 6)
        avg_countHelper = avg_countHelper + 6;
        avgT_0 = 0;
        avgT_0 = mean(avg_readings);
        avg_readings = 0;
        if(firstRun)
            avgT_2 = avgT_0;
            firstRun = false;
        end
        if (currentTime == 2)
            if(~enterLoop2)
                if(abs(avgT_0 - avgT_2) > thetaT)
                    if(num1 == 0)
                        avgT_5 = avgT_0;
                    end
                    if(num1 == N0)
                        pstart = avgT_5;
                        enterLoop2 = true;
                        output_file(i,2) = pstart;
                        num1 = 0;
                    else
                        num1 = num1 +1;
                    end
                else
                    output_file(i,4) = abs(avgT_0 - avgT_2);
                    output_file(i,5) = 10;
                    num1 = 0;
                end
                else
                    if(abs(avgT_0 - avgT_2) < thetaT)
                        if(num2 ==0)
                            avgT_5 = avgT_0;
                        end
                        if (num2 == N1)
                            pend = avgT_5;
                            enterLoop2 = false;
                            num2 = 0;
                            output_file(i,3) = pend;
                            output_file(i,4) = abs(avgT_0 - avgT_2);
                            allFloors(prevI:i) = currentFloor;
                            currentFloor = currentFloor + round((1/h0) .* eqConst .* (1+segma * temperature) * log10(pstart/pend))
                            output_file(i,6) = currentFloor;
                            prevI = i;
                        else
                            num2 = num2 +1;
                        end
                    else
                        output_file(i,4) = abs(avgT_0 - avgT_2);
                        output_file(i,5) = 20;
                        num2 = 0;
                    end
                end
                avgT_2 = avgT_0;
                currentTime = 0;
        end
            currentTime = currentTime + 1;
    end
end
time_sample = 1: length(input_data);
allFloors(prevI:i) = currentFloor;
t = tiledlayout(2,1)
lines_x = [127 246 378 534 687 821 977 1108];
ax1 = nexttile;
plot(time_sample, input_data)
ylabel('Pressure[hPa]')
% for i = 1:length(lines_x)
%     xline(lines_x(i),'r')
% end
nexttile
plot(time_sample, allFloors)
ylabel('Floor Number')

title(t,'Floor Detection')
xlabel(t,'Sample')

% Move plots closer together
xticklabels(ax1,{})
t.TileSpacing = 'compact';                    